### ISSUE TITLE
New pseudo type for ReQL queries in ReQL

### ISSUE THREAD

**Issue by gchpaco** [rethinkdb/rethinkdb#3636]
For purposes such as the `jobs` table (#3115, #3231) and a slow query log (#2238), it's necessary to display ReQL queries in a form comprehensible to the end user.  This is tricky because of the official drivers no two agree on nontrivial syntax, and they all contain some sort of kludge (keyword arguments, camel casing, operator overloading in JS; the limitations of `lambda` in Python; and Ruby of course is written in Ruby and the block syntax doesn't translate well).

I propose here a new pseudo-type `REQL_QUERY`, whose payload is the JSON form of the original query.  The drivers can then translate the query into whatever is appropriate for their target language.  This would also help things like the web UI make better HTML-ified query printing, since the whole AST is available.

At this time I do not want to add `r.eval()` per #1863, although that's an obvious extension after the pseudo-type exists.

**Comment by mlucy** [comment]
I would be OK with this.  The pseudotype should probably contain the ReQL version as well as the JSON representation.

One question is third-party clients: a lot of them don't have pretty-printing (and it's a lot of work to add), but we'd still like to let users of those clients use the jobs table.  An option that's more work for us is to have a runtime option drivers can provide to indicate whether or not they'd like `REQL_QUERY` pseudotypes to contain a pre-pretty-printed source representation (probably in JS since that's what the web UI uses).  The default would be to send the pretty-printed code as part of the pseudotype, so that users of third-party drivers which haven't been updated can see what they need to see, and our official drivers (plus any third party drivers that are smart enough) would opt out of the pre-pretty-printing and pretty-print themselves.

**Comment by gchpaco** [comment]
Sure, I have no objection to that.

**Comment by danielmewes** [comment]
For the record: I think it would be ok to always send the pretty-printed string alongside the query object. I don't think there's a need for an opt arg, since this is likely not performance critical.

**Comment by mlucy** [comment]
I think we all basically agree on this.  I'd be OK always including the pretty-printed string.  (Also, I think we already talked about this, but the `REQL_QUERY` pseudotype should contain the ReQL version as well as the JSON representation of the query.

**Comment by marshall007** [comment]
Once this is implemented, would it be possible to retrieve the function definition of a secondary index as `REQL_QUERY`? Also will I be able to do `query.coerceTo('REQL_QUERY')`?

It would be really awesome if we could compare existing secondary indices to what we expect them to be on app initialization so that we know whether they need to be dropped and recreated or not.

**Comment by danielmewes** [comment]
@marshall007 We actually have https://github.com/rethinkdb/rethinkdb/issues/3231 for the secondary index case, which I imagine we would base on this once we have it.
Turning a given query into a `REQL_QUERY` object is probably something we're going to add as well, but I would like to keep it out of this issue (and it won't make it into 2.0).

**Comment by mlucy** [comment]
I just talked about this with @coffeemug, and he sort of convinced me that this might not be worth the development effort right now.  Since the current state of this proposal involves writing a generic pretty-printer *anyway*, I think maybe we should just omit the pseudotype and put the pretty-printed form of the query in the jobs table for now.  If people complain, we can always add the pseudotype later, and the pretty-printing code won't have gone to waste or anything.

**Comment by gchpaco** [comment]
> think maybe we should just omit the pseudotype and put the pretty-printed form of the query in the jobs table for now

Which syntax should it use?  We have three official drivers each of which is *very* different from each other and each of which has some annoying misfeature (e.g. keyword arguments in JS and `lambda` in Python).  That was the whole reason to have the type in the first place, i.e. there *is* no sane common ground.  You yourself have complained that it's easier to just specify all three versions than fight with our polyglot tests.

Also I have written pretty printers before and I assure you it is easy.  Doing it four times is not meaningfully more difficult.

**Comment by mlucy** [comment]
I would just do it in JS since that's what the web UI uses.  (Or, alternatively, some sort of pseudocode.)  It obviously isn't perfect, but it's strictly less work than adding server-side pretty printing for third-party drivers + a pseudotype + support for that pseudotype in the official drivers.  (That said, it isn't *that* much work to have all those things.)

**Comment by coffeemug** [comment]
> Which syntax should it use?

I don't think it matters (but I'd opt for JavaScript). A database might run queries from clients written in completely different languages, and the person looking at the `jobs` table may or may not be familiar with those languages. I'm willing to bet that if we document that the `jobs` table uses a JavaScript notation people will say "oh, ok" and carry on with their work. Adding a pseudotype and printing the query in a specific language doesn't meaningfully improve the user experience IMO and doesn't seem worth doing.

**Comment by danielmewes** [comment]
I think in the long term the pseudo-type solution is strictly better. It will also make things like exporting secondary index functions much nicer, and allow us to add `r.eval` and related features.

It's definitely fine to skip the pseudo type for now and only do the JavaScript pretty-printer on the server. Since we need the pretty printer anyway, we will be able to still use it once we introduce the pseudo-type.

**Comment by gchpaco** [comment]
I kinda want to get the pseudo type in to 2.0 for driver developers even if we don't do anything with it (i.e. just print the server generated string).

As to pretty printing, there's some interesting divergences on what people actually *mean* by pretty printing, but what I mean is well supported by one of [Oleg Kiselyov's papers](http://okmij.org/ftp/continuations/PPYield/yield-pp.pdf).  Oppen's original optimal algorithm is imperative (which we don't care about) and uses coroutines (which is a bit awkward in many driver languages); Kiselyov's modification uses generators which are easy to retrofit even into C++, and it's performant, and it generates understandable results.

I tend to be pretty firm about needing to do something like this because I personally cannot read code easily without newline breaks.

I'm working on prototyping it now.

**Comment by coffeemug** [comment]
If we do introduce a pseudotype in 2.0, I think we should also add an optarg to `run` to control the results:

```py
r.db('rethinkdb').table('jobs').run(query_format='string|native|raw')
```

- `'raw'` would tell the driver to print the pseudotype json
- `'native'` would tell the driver to pretty-print the query locally
- `'string'` would tell the server to send a pretty-printed string

This way people who use drivers that don't support the pseudotype can still get a reasonable pretty-printed string. I'd also make the default be `'string'` because I think it's the most useful option in most cases and would play well with third-party drivers out of the box.

**Comment by mlucy** [comment]
@coffeemug -- we'd probably include the string in the pseudotype if we did it that way, so drivers that don't support it would still see the pretty-printed string when they printed the object.

@gchpaco -- I can see an argument for trying to make sure that we have all of our pseudotypes in for 2.0, but I really don't think this is worth the development effort right now.

**Comment by danielmewes** [comment]
After sleeping over this I think I agree with @mlucy and @coffeemug that we should just send a string in 2.0.

Having the full pseudo-type probably doesn't add that much extra work, but it's still going to be a couple hours of development and testing time. At the same time I really really doubt that third-party drivers are going to write their own pretty printers any time soon, at least not as long as we're only using the pseudo-type for the `jobs` table.

One thing we *should* do though is to strip large arrays, strings and binary values in the pretty-printed queries. If a user has a batched insert query running that inserts 10,000 documents each with a handful of fields, we don't want to pretty print all that data (at least not for the `jobs` table). Instead we should cut those off after a certain length, and insert a special marking (e.g. "...") into the pretty-printed query instead. (This by the way is another thing that would require a bit of extra thought and work for the full pseudo type, since we would probably need to represent the fact that a datum has been cut off somehow.)

**Comment by mlucy** [comment]
Planning to mark this as settled on Monday with the reduced scope for 2.0.

**Comment by gchpaco** [comment]
> One thing we should do though is to strip large arrays, strings and binary values in the pretty-printed queries.

The architecture I have for the pretty printer prototype will accommodate this easily, FWIW.
